### ISSUE TITLE
issue about the block mask

### ISSUE THREAD

**Issue by chengshuang18** [kuleshov-group/bd3lms#23]
Dear authors,

Thank you for the impressive work and open source. I have following questions.

### Problem Description
In `bd3lms/models/dit.py`, between line 696 and 707, you have this function:

```python
def gen_mask(self, seqlen, block_size, attn_backend='sdpa'):
    """Generates attention mask"""
    if attn_backend == 'flex' and FLEX_ATTN_AVAILABLE:
        self.block_diff_mask = create_block_mask(
            partial(block_diff_mask, block_size=block_size, n=seqlen),
            B=None, H=None, Q_LEN=seqlen*2, KV_LEN=seqlen*2)
    elif attn_backend == 'sdpa':
        self.block_diff_mask = block_diff_mask(
            b=None, h=None, q_idx=torch.arange(seqlen*2)[:, None], 
            kv_idx=torch.arange(seqlen*2)[None, :], block_size=block_size, n=seqlen)
    else:
        raise ValueError('Unknown attention backend')
```

We noticed two distinct issues with the above implementation:

### Issue 1: Incorrect or Missing `BLOCK_SIZE` Argument
In the branch `attn_backend == 'flex'`, the function `create_block_mask` is invoked without explicitly passing the necessary `BLOCK_SIZE` parameter. Due to this omission, it defaults to a BLOCK_SIZE of 128, causing incorrect mask creation. (This issue can be observed explicitly if one converts the generated mask to dense format using `.to_dense()`). This results in incorrect attention masks, severely impacting the correctness of computational results.

Therefore, we wonder if the published experimental results in the paper associated with this repository were generated using another correct version of the code or exactly this publicly available implementation?

### Issue 2: Handling of short sequence lengths in `create_block_mask`
Another issue is that `create_block_mask` appears not to handle short sequences (specifically sequence length smaller than 128). Is it correct that this function does not support `Q_LEN` less than 128, or is there some misunderstanding here?

### Software Versions:
- PyTorch version: 2.5.1
Could the above issues possibly relate to this PyTorch version?

**Comment by mariannearr** [comment]
Hello, I hope the below clarifies your concerns.

> Issue 1: Incorrect or Missing BLOCK_SIZE Argument

The [below line](https://github.com/kuleshov-group/bd3lms/blob/add6a518e1d0724d6c07fd0fe23c6708c5711cd2/models/dit.py#L709) in `dit.py` will create a block-causal mask (Fig 3 in our paper), where `block_size` is determined using `config.block_size`. 
```
self.block_diff_mask = create_block_mask(
        partial(block_diff_mask, block_size=block_size, n=seqlen),
        B=None, H=None, Q_LEN=seqlen*2, KV_LEN=seqlen*2)
```
To clarify, `BLOCK_SIZE` as used in FlexAttention refers to the size of the sub-matrices in the attention block, as discussed on the [FlexAttention blog post](https://pytorch.org/blog/flexattention/)

Can you please share your the scripts/configs used to produce your results? I have verified on my end that the generated attention masks are correct. 

> Issue 2: Handling of short sequence lengths in create_block_mask

The context length `config.model.length` should be greater than the block size specified by `config.block_size`. Can you share more details on the configs you use?

**Comment by mariannearr** [comment]
Please let me know if you have further questions. If the above resolves your concerns, please feel free to close this GitHub issue.

**Comment by mariannearr** [comment]
Closing this issue for now, feel free to reopen if you have further concerns!
